File: octave/ch1/Ex127.m
function Ex127
  test_case(10);
  test_case(11);
  test_case(17);
  test_case(561);
  test_case(1729);
  printf("done\n");
endfunction

function VAL = test_case(n)
  results = [prime(n), fast_prime(n, 128), carmichael_num(n)];
  printf("%d \t  Prime: %d \t Fast Prime: %d \t Carmichael: %d \n", [n, results]);
endfunction

function VAL = find_divisor(n, test)
  divides = @(a, b) rem(b, a) == 0;
  
  if test * test > n
    VAL = n;
  elseif divides(test, n)
    VAL = test;
  else
    VAL = find_divisor(n, test + 1);
  endif
endfunction

function VAL = smallest_divisor(n) 
  VAL = find_divisor(n, 2);
endfunction

function VAL = prime(n)
  VAL = n == smallest_divisor(n);
  ## VAL = fast_prime(n, 3);
endfunction

function VAL = eq_expmod(base, exp)
  VAL = expmod(base, exp, exp) == base;
endfunction

# recusrion depth problem with recursive call
# replace with iterative solution
function VAL = carmichael_num(n)
  try_it = @(a) expmod(a, n, n) == a;
  if prime(n)
    VAL = 0;
  else
    for i = n-1 : -1 : 0
      if i == 0
        VAL = 1;
        return;
      elseif  !try_it(i)
        VAL = 0;
        return;
      end
    endfor
  end
endfunction

function VAL = fermat_test(n)
  try_it = @(a) expmod(a, n, n) == a;
  VAL = try_it(1 + random(n-1));
endfunction

function VAL = fast_prime(n, times)
  if times == 0
    VAL = 1;
  elseif fermat_test(n)
    VAL = fast_prime(n, times-1);;
  else
    VAL = 0;
  endif
endfunction

function VAL = expmod(base, exp, m)
  even = @(n) rem(n, 2) == 0;
  square = @(n) n * n;
  if exp == 0
    VAL = 1;
  elseif even(exp)
    VAL = rem(square(expmod(base, exp/2, m)), m);
  else
    VAL = rem(base * expmod(base, exp-1, m), m);
  endif
endfunction

function VAL = random(limit)
  VAL = randi([0,limit]);
endfunction

File: rlang/ch1/ex127.R
# setwd("c:/Users/Yohan/IdeaProjects/sicp/rlang/ch1")
# getwd()
# load with source("filename")


square <- function(n) {
  n * n
}

random <- function(limit) {
  sample(0:limit, 1)
}

divides <- function(a, b) {
  0 == b %% a
}

find_divisor <- function(n, test) {
  if (test * test > n) {
    n
  } else if (divides(test, n)) {
    test
  } else {
    find_divisor(n, test + 1)
  }
}

smallest_divisor <- function(n) {
  find_divisor(n, 2)
}

is_prime <- function(n) {
  n == smallest_divisor(n)
}

fast_prime <- function(n, times) {
  if (times == 0) {
    TRUE
  } else if (fermat_test(n)) {
    fast_prime(n, times-1)
  } else {
    FALSE
  }
}

fermat_test <- function(n) {
  try_it <- function(a) { a == expmod(a, n, n) }
  try_it(1 + random(n-1))
}

expmod <- function(base, exp, m) {
  if (exp == 0) {
    1
  } else if (divides(2,exp)) {
    square(expmod(base, exp/2, m)) %% m
  } else {
    (base * expmod(base, exp-1, m)) %% m
  }
}

# cannot handle deep recursion
carmichael_num <- function(n) {
  try_it <- function(a) { a == expmod(a, n, n) }
  iter <- function(x) {
      for (i in x:0) {
          if (i == 0) {
              return(TRUE)
          } else if (!try_it(i)) {
              return(FALSE)
          }    
      }
  }    
  !is_prime(n) && iter(n-1)
}

test_number <- function(n) {
  print(noquote(paste(toString(n),
                      " Prime:", is_prime(n),
                      " Fast Prime:", fast_prime(n, 4),
                      " Carmichael:", carmichael_num(n))))
}


# Run in console
# > library(testthat)
# > test_file('this_file.R')

# test_that("test1", {
#   expect_equal(smallest_divisor(199), 199)
#   expect_equal(smallest_divisor(1999), 1999)
#   expect_equal(smallest_divisor(19999), 7)
# })

# Try these tests
test_number(9)
test_number(10)
test_number(11)
test_number(31)
test_number(59)
test_number(561)
test_number(1105)
test_number(1729)
test_number(2465)
test_number(2821)
test_number(6601)
